design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,529 @@
+`client.rbd-mirror-peer` user and key. Therefore, the singular RBD mirror key is rotated at the
+CephCluster level (above), but Rook should still update CephBlockPool statuses to identify when the
+peering token has been updated to use the latest peer token.
+
+For token updates, the CephBlockPool controller should reconcile when the parent CephCluster's
+`status.cephx.rbdMirrorPeer` is updated. When the CephBlockPool controller reconciles and creates
+its bootstrap token, it should copy `CephCluster.status.cephx.rbdMirrorPeer` to its own
+`status.cephx.peerToken`, which will indicate that the token has been updated with the latest key
+after a CephCluster key rotation event.
+
+```yaml
+kind: CephBlockPool
+# ...
+status:
+  # ...
+  cephx:
+    peerToken:
+      keyGeneration: 2
+      keyCephVersion: "20.2.2"
+```
+
+### CephFilesystem mirror example
+
+TODO: need design for this
+
+### CephClient example
+
+CephClient stands out due to its simplicity. It will have one and exactly one key which can serve
+any purpose (neither daemon key nor peer token). The interfaces are simplified to reflect this.
+
+```yaml
+kind: CephClient
+# ...
+spec:
+  # ...
+  security:
+    cephx:
+      keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+      keyGeneration: <int> # used with KeyGeneration
+    # (room for future spec.security options)
+  # ...
+status:
+  cephx:
+    keyGeneration: 1 # e.g.
+    keyCephVersion: "20.2.0" # e.g.
+```
+
+## Dependencies
+
+Ceph Tentacle (v20) provides a new `ceph auth rotate` command merged here:
+https://github.com/ceph/ceph/pull/58121. Rook will rely on this command to rotate keys.
+
+A few CephX technical details that are important for understanding CephX key rotation development
+are summarized below. More details here: https://docs.ceph.com/en/latest/dev/cephx/
+
+CephX keys minted by Rook are only used by Ceph for initial daemon connection. Internally, all Ceph
+connections use "service" keys with laddered expiration times.
+
+By default, Ceph service keys allow at least 2 hours from key rotation until the client must be
+updated with the new key. There are 3 service keys with TTLs 1 hour, 2 hours, and 3 hours from the
+time when service keys were last refreshed. Assuming the first TTL expiration is imminent, it is
+still at least 2 hours until the 3-hour TTL expiration.
+
+Ceph does not allow keyrings to contain multiple keys for a given client/daemon. When a key is
+rotated, the old key is removed and new key replaces it with no ability to have a 2-valid-keys
+transition period.
+
+Ceph's `auth_service_ticket_ttl` and `auth_mon_ticket_ttl` config options allow users to
+shorten/lengthen that time as desired/needed. Note that this isn't immediate: it will take an
+unknown amount of time for internal service keys to update.
+
+Ceph config `debug_auth=30` provides maximum CephX debug logs, as a development aid.
+
+## Ceph admin key rotation
```

Comment:
@travisn after our chat earlier, I sat down to think through all the ways admin key rotation could fail. There is nonzero risk of bricking a CephCluster, so I wanted to plan out the pseudocode and write it down for review.

On Monday, I will begin the code work on this, and I will adjust as needed based on any feedback here.